File: IPWorks Samples/WebSocket Server/net/wsserver.cs
/*
 * IPWorks 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using nsoftware.IPWorks;

class wsserverDemo
{
  private static WSServer wsserver;

  private static void wsserver_OnConnected(object sender, WSServerConnectedEventArgs e)
  {
    Console.WriteLine(wsserver.Connections[e.ConnectionId].RemoteHost + " has connected.");
  }

  private static void wsserver_OnDataIn(object sender, WSServerDataInEventArgs e)
  {
    Console.WriteLine("Received '" + e.Text + "'.");
  }

  private static void wsserver_OnDisconnected(object sender, WSServerDisconnectedEventArgs e)
  {
    Console.WriteLine(wsserver.Connections[e.ConnectionId].RemoteHost + " has disconnected - " + e.Description + ".");
  }

  private static void wsserver_OnError(object sender, WSServerErrorEventArgs e)
  {
    Console.WriteLine(e.Description);
  }

  static void Main(string[] args)
  {
    wsserver = new WSServer();

    if (args.Length < 1)
    {
      Console.WriteLine("usage: wsserver [options] port");
      Console.WriteLine("Options: ");
      Console.WriteLine("  -cert      the subject of a certificate in the user's certificate store to be used during SSL/TLS negotiation (default no SSL/TLS)");
      Console.WriteLine("  port       the TCP port in the local host where the component listens");
      Console.WriteLine("\r\nExample: wsserver -cert certSubject 4444");
    }
    else
    {
      wsserver.OnConnected += wsserver_OnConnected;
      wsserver.OnDataIn += wsserver_OnDataIn;
      wsserver.OnDisconnected += wsserver_OnDisconnected;
      wsserver.OnError += wsserver_OnError;

      try
      {
        // Parse arguments into component.
        wsserver.LocalPort = int.Parse(args[args.Length - 1]);

        for (int i = 0; i < args.Length; i++)
        {
          if (args[i].StartsWith("-"))
          {
            if (args[i].Equals("-cert"))
            {
              wsserver.SSLCert = new Certificate(CertStoreTypes.cstUser, "MY", "", args[i + 1]);  // args[i + 1] corresponds to the value of args[i]
              wsserver.UseSSL = true;
            }
          }
        }

        // Start lisenting for connections.
        wsserver.StartListening();

        // Process user commands.
        Console.WriteLine("Type \"?\" or \"help\" for a list of commands.");
        string command;
        string[] arguments;

        while (true)
        {
          command = Console.ReadLine();
          arguments = command.Split();

          if (arguments[0].Equals("?") || arguments[0].Equals("help"))
          {
            Console.WriteLine("Commands: ");
            Console.WriteLine("  ?                            display the list of valid commands");
            Console.WriteLine("  help                         display the list of valid commands");
            Console.WriteLine("  send <text>                  send data to connected clients");
            Console.WriteLine("  quit                         exit the application");
          }
          else if (arguments[0].Equals("send"))
          {
            if (arguments.Length > 1)
            {
              string textToSend = "";
              for (int i = 1; i < arguments.Length; i++)
              {
                if (i < arguments.Length - 1) textToSend += arguments[i] + " ";
                else textToSend += arguments[i];
              }
              foreach (WSConnection connection in wsserver.Connections.Values)
              {
                wsserver.SendText(connection.ConnectionId, textToSend);
              }
            }
            else
            {
              Console.WriteLine("Please supply the text that you would like to send.");
            }
          }
          else if (arguments[0].Equals("quit"))
          {
            wsserver.Shutdown();
            break;
          }
          else if (arguments[0].Equals(""))
          {
            // Do nothing.
          }
          else
          {
            Console.WriteLine("Invalid command.");
          }

          Console.Write("wsserver> ");
        }
      }
      catch (Exception e)
      {
        Console.WriteLine(e.Message);
      }
      Console.WriteLine("Press any key to exit...");
      Console.ReadKey();
    }
  }
}





class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add an key to the dictionary for each argument
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/" then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}

File: IPWorks Samples/WebSocket Server/c++/wsserver.cpp
/*
 * IPWorks 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#include "../../include/ipworks.h"
#define LINE_LEN 80

class MyWSServer : public WSServer
{
	int FireConnected(WSServerConnectedEventParams *e)
	{
		printf("%s has connected.\n", this->GetWSConnectionRemoteHost(e->ConnectionId));
		return 0;
	}

	int FireDisconnected(WSServerDisconnectedEventParams *e)
	{
		printf("Disconnected from %s.\n", this->GetWSConnectionRemoteHost(e->ConnectionId));
		return 0;
	}

	int FireDataIn(WSServerDataInEventParams *e)
	{
		printf("%s - Echoing '%s' back to client.\n", this->GetWSConnectionRemoteHost(e->ConnectionId), e->Text);
		char *text = (char*)e->Text;
		this->Send(e->ConnectionId, text, strlen(text));
		return 0;
	}
};

int main(int argc, char* argv[])
{
	MyWSServer wsserver;

	char buffer[LINE_LEN];

	printf("*****************************************************************\n");
	printf("* This demo shows how to set up an echo server using WSServer.  *\n");
	printf("*****************************************************************\n");

	printf("Local Port: ");
	fgets(buffer, LINE_LEN, stdin);
	buffer[strlen(buffer)-1] = '\0';
	wsserver.SetLocalPort(atoi(buffer));
	
	printf("Use SSL? [y/n]");
	char response[LINE_LEN];

	fgets(response,LINE_LEN,stdin);
	response[strlen(response)-1] = '\0';
	if (strcmp(response,"y") == 0)
		wsserver.SetUseSSL(true);
		
	if (wsserver.GetUseSSL())
	{
		//This demo will assume a PFX file is being used for the server certificate but other formats can be used.
		wsserver.SetSSLCertStoreType(CST_PFXFILE);

		printf("PFX File to use for server certificate: ");
		fgets(buffer,LINE_LEN,stdin);
		buffer[strlen(buffer)-1] = '\0';
		wsserver.SetSSLCertStore(buffer,strlen(buffer));

		printf("PFX File Password: ");
		fgets(buffer,LINE_LEN,stdin);
		buffer[strlen(buffer)-1] = '\0';
		wsserver.SetSSLCertStorePassword(buffer);

		//The default value of "*" picks the first private key in the certificate. For simplicity this demo will use that value.
		wsserver.SetSSLCertSubject("*");
	}

	int ret_code = wsserver.StartListening();

	if(ret_code)
	{
		printf("Error: %i - %s\n", ret_code,wsserver.GetLastError());
		goto done;
	}

	printf("Listening...\n");

	while(true)
	{
		wsserver.DoEvents();
	}

done:
	if (wsserver.GetListening())
	{
		wsserver.StopListening();
	}

	printf( "Exiting... (press enter)\n" );
	getchar();

	return 0;
}

